hibernate-core/src/main/java/org/hibernate/event/internal/AbstractSaveEventListener.java

```diff
@@ -269,7 +269,8 @@ protected Serializable performSaveOrReplicate(
 		boolean substitute = substituteValuesIfNecessary( entity, id, values, persister, source );
 
 		if ( persister.hasCollections() ) {
-			substitute = substitute || visitCollectionsBeforeSave( entity, id, values, types, source );
+			final boolean substituteBecauseOfCollections = visitCollectionsBeforeSave( entity, id, values, types, source );
+			substitute = substitute || substituteBecauseOfCollections;
```

Comment:
`visitCollectionsBeforeSave()` is a method with side effect. The side effect would not happen if `substitute` has been true (e.g., versioned entity) due to boolean operator short-circuiting.